src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+  srfft_ = new SplitRadixRealFft<BaseFloat>(fft_length);
+
+  Vector<float> filter_padded(fft_length);
+  filter_padded.Range(0, filter_length).CopyFromVec(filter);
+  srfft_->Compute(filter_padded.Data(), true);
+
+  Vector<float> temp_pad(filter_length - 1);
+  temp_pad.SetZero();
+  Vector<float> signal_block_padded(fft_length);
+
+  for (int po = 0; po < signal_length; po += block_length) {
+    // get a block of the signal
+    int process_length = block_length;
+    if (signal_length - po < process_length) process_length = signal_length - po;
+    signal_block_padded.SetZero();
+    signal_block_padded.Range(0, process_length).CopyFromVec(signal->Range(po, process_length));
+
+    srfft_->Compute(signal_block_padded.Data(), true);
+
+    ElementwiseProductOfFft(filter_padded, &signal_block_padded);
+
+    srfft_->Compute(signal_block_padded.Data(), false);
+    signal_block_padded.Scale(1.0 / fft_length);
+
+    // combine the block
+    if (po + block_length < signal_length) {       // current block is not the last block
+      signal->Range(po, block_length).CopyFromVec(signal_block_padded.Range(0, block_length));
+      signal->Range(po, filter_length - 1).AddVec(1.0, temp_pad);
+      temp_pad.CopyFromVec(signal_block_padded.Range(block_length, filter_length - 1));
+    } else {
+      signal->Range(po, signal_length - po).CopyFromVec(
+                        signal_block_padded.Range(0, signal_length - po));
+      if (filter_length - 1 < signal_length - po)
+        signal->Range(po, filter_length - 1).AddVec(1.0, temp_pad);
+      else
+        signal->Range(po, signal_length - po).AddVec(1.0, temp_pad.Range(0, signal_length - po));
+    }
+  }
+}
+
+void AddVectorsOfUnequalLength(const Vector<BaseFloat> &filter, Vector<BaseFloat> *signal) {
+  for (int po = 0; po < signal->Dim(); po += filter.Dim()) {
+    int block_length = filter.Dim();
+    if (signal->Dim() - po < block_length) block_length = signal->Dim() - po;
+    signal->Range(po, block_length).AddVec(1.0, filter.Range(0, block_length));
+  }
+}
+
+BaseFloat MaxAbsolute(const Vector<BaseFloat> &vector) {
+  return std::max(std::abs(vector.Max()), std::abs(vector.Min()));
+}
+
+BaseFloat ComputeEarlyReverbEnergy(const Vector<BaseFloat> &rir, const Vector<BaseFloat> &signal,
+                                   BaseFloat samp_freq) {
+      int peak_index = 0;
+      rir.Max(&peak_index);
+      KALDI_VLOG(1) << "peak index is " << peak_index;
+
+      const float sec_before_peak = 0.001;
+      const float sec_after_peak = 0.05;
+      int early_rir_start_index = peak_index - sec_before_peak * samp_freq;
+      int early_rir_end_index = peak_index + sec_after_peak * samp_freq;
+      if (early_rir_start_index < 0) early_rir_start_index = 0;
+      if (early_rir_end_index > rir.Dim()) early_rir_end_index = rir.Dim();
+
+      int duration = early_rir_end_index - early_rir_start_index;
+      Vector<BaseFloat> early_rir(rir.Range(early_rir_start_index, duration));
```

Comment:
This is probably not a big deal because of a small size, but if `BlockConvolveSignals()` could take the first arg as a `VectorBase<>`, the matrix copy would not be necessary here, you could use the result of `.Range()` directly. Not sure if `VectorBase<>` works there, just a suggestion.